### DISCUSSION TITLE
MediaMTX + SSL

### DISCUSSION THREAD

**Discussion by ringtailedfox** [bluenviron/mediamtx#3322] [Questions]
### Question

I have used Let's Encrypt and CertBot to generate valid SSL certificates for Apache on a test server.. is there a way for me to get MediaMTX to use these as well?  So far, it seems to just say 'authentication failed" no matter what I do if i'm not connecting via localhost or LAN IP address (those two work)...

Ideally, I'd love to be able to set up a stream or playlist on a webpage with valid SSL inherited from Apache2

**Comment by Taps7734** [comment]
I have an IIS web server running on the same machine that uses ports 80 and 443 for the site.  I use Let's Encrypt to create certificates for that site using the Win-Acme solution, WACS.EXE.  You can find the password using (a)Manage Renewals > (d)Show Details > .pfx Password

Those certificates and the keys are then saved in the ProgramData folder. This is from memory, so I cannot tell you the path in linux servers.  Which I am assuming you are using, but you didnt specify. I do not believe the *nix programs use a PFX, but possibly a PEM. But the process should be similar.

You can use OpenSSL to export the cert and unencrypted key and save them to the MediaMTX folder so they can be used. I think the issue is going to be the UNENCRYPTED key for you.

I use the following to get the unencrypted key : 
`openssl pkcs12 -in server.pfx -passin pass:PFXPasswordFromWACS -nokeys -out server.crt`

`openssl pkcs12 -in server.pfx -nocerts -out server.key -passin pass:PFXPasswordFromWACS -passout pass:WhateverPasswordYouChoose`

`openssl rsa -in server.key -out server-unencrypt.key -passin pass:WhateverPasswordYouChoose`

Then I point MediaMTX to use the server-unencrypt.key and server.crt.

**Comment by ringtailedfox** [comment]
I found the sections within mediamtx.yml that asked for .pem files, so extraction was thankfully unnecessary.  i now have functional HTTPS/SSL for MediaMTX!  I'll keep your instructions as a standby, just in case this no longer works. :D

**Reply by vdiaz3** [reply]
awesome! please can you share how do you enable https? thanks a loT!!!

**Reply by ringtailedfox** [reply]
in your mediamtx.yml configuration file, look for these lines:

```
hlsServerKey: /opt/mediamtx/privkey.pem
# Path to the server certificate.
hlsServerCert: /opt/mediamtx/fullchain.pem
```
copy your .pem-encrypted keys to say, /opt/mediamtx/ (where i have my install), and it should work!

i copied my .pem keys from /etc/letsencrypt/archive/www.website.whatever/ to /opt/mediamtx/ just to ensure there were no permissions issues...

as for RTMPS and RTSPS... i have no idea how to do those, but HLS works with HTTPS now

[a comment by an automated account is not shown]
